cmd/zilstat.in

```diff
@@ -0,0 +1,470 @@
+import os
+import re
+
+from collections import defaultdict
+from signal import signal, SIGINT, SIG_DFL
+
+cols = {
+	# hdr:       [size,      scale, 	 kstat name]
+	"time":      [8,         -1,         "time"],
+	"pool":      [12,        -1,         "pool"],
+	"ds":        [12,        -1,         "dataset_name"],
+	"obj":       [12,        -1,         "objset"],
+	"zcc":       [10,        1000,       "zil_commit_count"],
+	"zcwc":      [10,        1000,       "zil_commit_writer_count"],
+	"ziic":      [10,        1000,       "zil_itx_indirect_count"],
+	"zic":       [10,        1000,       "zil_itx_count"],
+	"ziib":      [10,        1024,       "zil_itx_indirect_bytes"],
+	"zicc":      [10,        1000,       "zil_itx_copied_count"],
+	"zicb":      [10,        1024,       "zil_itx_copied_bytes"],
+	"zinc":      [10,        1000,       "zil_itx_needcopy_count"],
+	"zinb":      [10,        1024,       "zil_itx_needcopy_bytes"],
+	"zimnc":     [10,        1000,       "zil_itx_metaslab_normal_count"],
+	"zimnb":     [10,        1024,       "zil_itx_metaslab_normal_bytes"],
+	"zimsc":     [10,        1000,       "zil_itx_metaslab_slog_count"],
+	"zimsb":     [10,        1024,       "zil_itx_metaslab_slog_bytes"],
+}
+
+hdr = ["time", "pool", "ds", "obj", "zcc", "zcwc", "ziic", "zic", "ziib", \
+	"zicc", "zicb", "zinc", "zinb", "zimnc", "zimnb", "zimsc", "zimsb"]
+
+ghdr = ["time", "zcc", "zcwc", "ziic", "zic", "ziib", "zicc", "zicb",
+	"zinc", "zinb", "zimnc", "zimnb", "zimsc", "zimsb"]
+
+cmd = ("Usage: zilstat [-hgdv] [-i interval] [-p pool_name]")
+
+curr = {}
+diff = {}
+kstat = {}
+ds_pairs = {}
+pool_name = None
+dataset_name = None
+interval = 0
+sep = "  "
+gFlag = True
+dsFlag = False
+
+def prettynum(sz, scale, num=0):
+	suffix = [' ', 'K', 'M', 'G', 'T', 'P', 'E', 'Z']
+	index = 0
+	save = 0
+
+	if scale == -1:
+		return "%*s" % (sz, num)
+
+	# Rounding error, return 0
+	elif 0 < num < 1:
+		num = 0
+
+	while num > scale and index < 5:
+		save = num
+		num = num / scale
+		index += 1
+
+	if index == 0:
+		return "%*d" % (sz, num)
+
+	if (save / scale) < 10:
+		return "%*.1f%s" % (sz - 1, num, suffix[index])
+	else:
+		return "%*d%s" % (sz - 1, num, suffix[index])
+
+def print_header():
+	global hdr
+	global sep
+	try:
+		for col in hdr:
+			new_col = col
+			if interval > 0 and col not in ['time', 'pool', 'ds', 'obj']:
+				new_col += "/s"
+			sys.stdout.write("%*s%s" % (cols[col][0], new_col, sep))
+		sys.stdout.write("\n")
+	except IOError as e:
+		if e.errno == errno.EPIPE:
+			sys.exit(1)
+
+def print_values(v):
+	global hdr
+	global sep
+	try:
+		for col in hdr:
+			val = v[cols[col][2]]
+			if col not in ['time', 'pool', 'ds', 'obj'] and interval > 0:
+				val = v[cols[col][2]] // interval
+			sys.stdout.write("%s%s" % (
+			    prettynum(cols[col][0], cols[col][1], val), sep))
+		sys.stdout.write("\n")
+	except IOError as e:
+		if e.errno == errno.EPIPE:
+			sys.exit(1)
+
+def print_dict(d):
+	try:
+		for pool in list(d.keys()):
```

Comment:
isn't this the same as the easier-to-read
`for pool in d:`
?

Reply:
I think it's a little different as `d` is a multi-dimensional diction here, e.g., d[pool][objid][zilstats].

Reply:
The suggestion is correct:
```python
>>> d = dict(pool0=dict(obj0=dict(stat0=1,stat1=2),obj1=dict(stat0=3,stat1=4)),pool1=dict(obj0=dict(stat0=5,stat1=6)))
>>> d.keys()
dict_keys(['pool0', 'pool1'])
>>> for k in d: print(k)
...
pool0
pool1
>>> for pool in d:
...     for objset in d[pool]:
...             print(d[pool][objset])
...
{'stat0': 1, 'stat1': 2}
{'stat0': 3, 'stat1': 4}
{'stat0': 5, 'stat1': 6}
```

Reply:
Thank you @freqlabs for clarification. I have updated according to @richardelling suggestion.